src/routing/orshttp_wrapper.cpp

```diff
@@ -0,0 +1,198 @@
+
+  return query;
+}
+
+std::string OrsHttpWrapper::send_then_receive(std::string query) const {
+  std::string response;
+
+  try {
+    boost::asio::io_service io_service;
+
+    tcp::resolver r(io_service);
+
+    tcp::resolver::query q(_server.host, _server.port);
+
+    tcp::socket s(io_service);
+    boost::asio::connect(s, r.resolve(q));
+
+    boost::asio::write(s, boost::asio::buffer(query));
+
+    char buf[512];
+    boost::system::error_code error;
+    for (;;) {
+      std::size_t len = s.read_some(boost::asio::buffer(buf), error);
+      response.append(buf, len);
+      if (error == boost::asio::error::eof) {
+        // Connection closed cleanly.
+        break;
+      } else {
+        if (error) {
+          throw boost::system::system_error(error);
+        }
+      }
+    }
+  } catch (boost::system::system_error& e) {
+    throw Exception("Failed to connect to ORS at " + _server.host + ":" +
+                    _server.port);
+  }
+  return response;
+}
+
+Matrix<Cost>
+OrsHttpWrapper::get_matrix(const std::vector<Location>& locs) const {
+  std::string query = this->build_query(locs, "matrix");
+
+  std::string response = this->send_then_receive(query);
+
+  // Removing headers.
+  std::string json_content = response.substr(response.find("{"));
+
+  // Expected matrix size.
+  std::size_t m_size = locs.size();
+
+  // Checking everything is fine in the response.
+  rapidjson::Document infos;
+#ifdef NDEBUG
+  infos.Parse(json_content.c_str());
+#else
+  assert(!infos.Parse(json_content.c_str()).HasParseError());
+#endif
+  if (infos.HasMember("error")) {
+    throw Exception("ORS matrix:" +
+                    std::string(infos["error"]["message"].GetString()));
+  }
+  assert(infos["durations"].Size() == m_size);
+
+  // Build matrix while checking for unfound routes to avoid
+  // unexpected behavior (ORS raises 'null').
+  Matrix<Cost> m(m_size);
+
+  std::vector<unsigned> nb_unfound_from_loc(m_size, 0);
+  std::vector<unsigned> nb_unfound_to_loc(m_size, 0);
+
+  for (rapidjson::SizeType i = 0; i < infos["durations"].Size(); ++i) {
+    const auto& line = infos["durations"][i];
+    assert(line.Size() == m_size);
+    for (rapidjson::SizeType j = 0; j < line.Size(); ++j) {
+      if (line[j].IsNull()) {
+        // No route found between i and j. Just storing info as we
+        // don't know yet which location is responsible between i
+        // and j.
+        ++nb_unfound_from_loc[i];
+        ++nb_unfound_to_loc[j];
+      } else {
+        auto cost = round_cost(line[j].GetDouble());
+        m[i][j] = cost;
+      }
+    }
+  }
+
+  check_unfound(locs, nb_unfound_from_loc, nb_unfound_to_loc);
+
+  return m;
+}
+
+void OrsHttpWrapper::add_route_info(Route& route) const {
+  // Ordering locations for the given steps.
+  std::vector<Location> ordered_locations;
+  for (const auto& step : route.steps) {
+    ordered_locations.push_back(step.location);
+  }
+
+  std::string extra_args = "";
```

Comment:
- Currently, ORS does not support alternatives. If we'll do so in the future, the option will most probably default to false in order to not break existing code.
- I am currently uncertain about what steps mean because the ORSM documentation is a bit sparse about it. I'll do a more thorough investigation of this issue soon.
- I can add the option `geometry_simplify=false` (which is the default)
- I can add the option `continue_straight=false` (which is the default)

Reply:
ORS has no option like `steps=false`to deliver distances and durations without instructions. I have written a feature request to include such an option.